Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,484 @@
+	}
+}
+
+void msblenMaterialsExportHelper::handlePassthrough(const Material* mat,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler,
+	bNode* sourceNode)
+{
+	auto imageInput = getInputSocket(sourceNode, colorIdentifier);
+	if (!imageInput)
+		return;
+
+	setValueFromSocket(mat,
+		imageInput, textureType,
+		resetIfInputIsTexture,
+		setColorHandler,
+		setTextureHandler);
+}
+
+void msblenMaterialsExportHelper::handleSocketValue(bNodeSocket* socket,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler)
+{
+	// The socket is using a direct value, not the output of another node.
+	// Clear any texture that was set:
+	if (setTextureHandler) {
+		setTextureHandler(ms::InvalidID);
+	}
+
+	if (setColorHandler) {
+		mu::float4 colorValue;
+
+		switch (socket->type)
+		{
+		case SOCK_RGBA: {
+			auto defaultValue = (bNodeSocketValueRGBA*)socket->default_value;
+			auto val = defaultValue->value;
+			colorValue = mu::float4{ val[0], val[1], val[2], val[3] };
+			break;
+		}
+		case SOCK_FLOAT: {
+			auto defaultValue = (bNodeSocketValueFloat*)socket->default_value;
+			auto val = defaultValue->value;
+			colorValue = mu::float4{ val, val, val, val };
+			break;
+		}
+		}
+		setColorHandler(colorValue);
+	}
+}
+
+void msblenMaterialsExportHelper::setValueFromSocket(const Material* mat,
+	bNodeSocket* socket,
+	ms::TextureType textureType,
+	bool resetIfInputIsTexture,
+	std::function<void(const mu::float4& colorValue)> setColorHandler,
+	std::function<void(int textureId)> setTextureHandler)
+{
+	// If there is nothing connected to the socket, send the value set on the socket:
+	if (!socket->link) {
+		handleSocketValue(socket, setColorHandler, setTextureHandler);
+		return;
+	}
+
+	// If there is an image linked to the socket, send that as a texture:
+	auto sourceNode = removeReroutes(socket->link->fromnode, mat);
+
+	if (sourceNode->flag & NODE_MUTED) {
+		return;
+	}
+
+	if (!m_settings->sync_textures) {
+		setTextureHandler = nullptr;
+	}
+
+	switch (sourceNode->type) {
+	case SH_NODE_TEX_IMAGE:
+	{
+		handleImageNode(textureType, resetIfInputIsTexture, setColorHandler, setTextureHandler, sourceNode);
+		break;
+	}
+	case SH_NODE_NORMAL_MAP:
+	{
+		handleNormalMapNode(mat, textureType, resetIfInputIsTexture, setColorHandler, setTextureHandler, sourceNode);
+		break;
+	}
+	case SH_NODE_DISPLACEMENT:
+	{
+		handleDisplacementNode(mat, textureType, resetIfInputIsTexture, setColorHandler, setTextureHandler, sourceNode);
+		break;
+	}
+	// Pass input straight through these:
+	case SH_NODE_GAMMA:
+	case SH_NODE_HUE_SAT:
+	{
+		handlePassthrough(mat, textureType, resetIfInputIsTexture, setColorHandler, setTextureHandler, sourceNode);
+		break;
+	}
+	}
+}
+
+void msblenMaterialsExportHelper::setShaderFromBSDF(ms::StandardMaterial& stdmat, bNode* bsdfNode)
```

Comment:
Is the standard material referring to the shader in Unity? Since we support multiple pipelines, the name can be misleading. Would something like PBRMaterial make more sense?

Reply:
This is what it was called in meshsync before. It refers to Unity's standard shader.

Reply:
My changes were additions to the existing class, other DCC tools that used this would not be affected and we can implement the additional properties on them too, no need to make a new class, I think.
Baking the maps on the DCC tool side would require knowledge about the pipeline of the server and separate implementations for each DCC tool, whereas baking them in Unity is done in a single place for its active pipeline, no matter what DCC tool the data came from so would have a lower maintenance cost.

Reply:
If that StandardMaterial is also being used for URP/HDRP, maybe we should rename it to something more generic, or create a new one as Sean mentioned, because it won't just be used for Unity's standard shader anymore

Reply:
> other DCC tools that used this would not be affected
 
 This is not a concern for the work you have done, but it should be for renaming the class, as I suggest
 

>  Baking the maps on the DCC tool side would require knowledge about the pipeline of the server and separate implementations for each DCC tool, whereas baking them in Unity is done in a single place for its active pipeline, no matter what DCC tool the data came from so would have a lower maintenance cost.

I agree, this is why we must not have classes on the client that are specific to the shader used in Unity

Reply:
I still don't see the value in renaming this. It just means it's a standard material with a standard pbr shader. If in the future we support special materials or shader effects these would be new classes with different maps and shaders.

Reply:
In this case, we should have a good xml comment clarifying that this is not referring to the Unity Standard Shader, its just a generic PBR material

Reply:
I've updated the comment at the top of the StandardMaterial class.